lib/SILOptimizer/LoopTransforms/COWArrayOpt.cpp

```diff
@@ -1470,49 +1038,62 @@ bool COWArrayOpt::hoistMakeMutable(ArraySemanticsCall MakeMutable) {
 
   // We can hoist address projections (even if they are only conditionally
   // executed).
-  auto ArrayAddrBase = stripUnaryAddressProjections(CurrentArrayAddr);
-  SILBasicBlock *ArrayAddrBaseBB = ArrayAddrBase->getParentBlock();
-
-  if (ArrayAddrBaseBB && !DomTree->dominates(ArrayAddrBaseBB, Preheader)) {
+  SILValue ArrayAddrBase = getArrayAddressBase(CurrentArrayAddr);
+  if (!ArrayAddrBase) {
     LLVM_DEBUG(llvm::dbgs() << "    Skipping Array: does not dominate loop!\n");
     return false;
   }
 
   SmallVector<unsigned, 4> AccessPath;
   SILValue ArrayContainer = getAccessPath(CurrentArrayAddr, AccessPath);
+  bool arrayContainerIsUnique = checkUniqueArrayContainer(ArrayContainer);
+
+  StructUseCollector StructUses;
+  StructUses.collectUses(ArrayContainer, AccessPath);
```

Comment:
The motivation was that it is now also needed in case of hasLoopOnlyDestructorSafeArrayOperations() == true.
But you are right, I can move it past the arrayContainerIsUnique check.